features/netsocket/Socket.h

```diff
@@ -21,93 +21,121 @@
 #define SOCKET_H
 
 #include "netsocket/SocketAddress.h"
-#include "netsocket/NetworkStack.h"
-#include "rtos/Mutex.h"
 #include "Callback.h"
-#include "mbed_toolchain.h"
-
 
 /** Abstract socket class
  */
 class Socket {
 public:
-    /** Destroy a socket
+    /** Destroy a socket.
      *
      *  Closes socket if the socket is still open
      */
     virtual ~Socket() {}
 
-    /** Opens a socket
+    /** Close the socket.
```

Comment:
I think listen() and accept() should go into the base level. They're as theoretically generic as all the other functions, and it would be harder to change later - particularly if you're trying to keep `Socket` totally abstract to the point of excluding default "unsupported" errors.

Reply:
Accept is a bit painfull because it does not return any abstract type, it needs a pointer of already allocated object which it then fills in.
`nsapi_error_t accept(TCPSocket *connection, SocketAddress *address = NULL);`

One option would be that we create new `accept()` call that is able to work with abstract types. After all, accepting from TCPserver is fine to return `Socket*` and just use it and call `close()` after you are done.

Reply:
Ah, right, I recall this discussion now. Any generic `accept` would need a different prototype, yes. So maybe define that now, and leave the other one deprecated in `TCPServer`?

Reply:
Generic `accept()` call now implemented